src/client/csfle/state_machine.rs

```diff
@@ -0,0 +1,171 @@
+                    }
+                    ctx.mongo_done()?;
+                }
+                State::NeedKms => {
+                    let ctx = result_mut(&mut ctx)?;
+                    let scope = ctx.kms_scope();
+                    let mut kms_ctxen: Vec<Result<_>> = vec![];
+                    while let Some(kms_ctx) = scope.next_kms_ctx() {
+                        kms_ctxen.push(Ok(kms_ctx));
+                    }
+                    stream::iter(kms_ctxen)
+                        .try_for_each_concurrent(None, |mut kms_ctx| async move {
+                            let endpoint = kms_ctx.endpoint()?;
+                            let addr = ServerAddress::parse(endpoint)?;
+                            let provider = kms_ctx.kms_provider()?;
+                            let tls_options = csfle
+                                .opts()
+                                .tls_options
+                                .as_ref()
+                                .and_then(|tls| tls.get(&provider))
+                                .cloned()
+                                .unwrap_or_default();
+                            let mut stream = AsyncStream::connect(
+                                StreamOptions::builder()
+                                    .address(addr)
+                                    .tls_options(tls_options)
+                                    .build(),
+                            )
+                            .await?;
+                            stream.write_all(kms_ctx.message()?).await?;
+                            let mut buf = vec![];
+                            while kms_ctx.bytes_needed() > 0 {
+                                let buf_size = kms_ctx.bytes_needed().try_into().map_err(|e| {
+                                    Error::internal(format!("buffer size overflow: {}", e))
+                                })?;
+                                buf.resize(buf_size, 0);
+                                let count = stream.read(&mut buf).await?;
+                                kms_ctx.feed(&buf[0..count])?;
+                            }
+                            Ok(())
+                        })
+                        .await?;
+                }
+                State::NeedKmsCredentials => todo!("RUST-1314"),
+                State::Ready => {
+                    let (tx, rx) = oneshot::channel();
+                    let mut thread_ctx = std::mem::replace(
+                        &mut ctx,
+                        Err(Error::internal("crypto context not present")),
+                    )?;
+                    csfle.crypto_threads.spawn(move || {
+                        let result = thread_ctx.finalize().map(|doc| doc.to_owned());
+                        let _ = tx.send((thread_ctx, result));
+                    });
+                    let (ctx_again, output) = rx
+                        .await
+                        .map_err(|_| Error::internal("crypto thread dropped"))?;
```

Comment:
what are the consequences of this happening? also are you aware of any ways this could reasonably occur?